#!/usr/bin/env python3

import os

os.environ["PATH"] = "/usr/local/bin:/usr/local/anaconda3/bin/:" + os.environ["PATH"]

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import argparse
import sys
import seaborn as sns
import string

def get_chrmaxes(chromlist, endlist):
    """Get the maximum position value in each chromosome"""
    chrmaxes = {}
    for i in zip(chromlist, endlist):
        if i[0] not in chrmaxes:
            chrmaxes[i[0]] = int(i[1])
        else:
            chrmaxes[i[0]] = max(int(i[1]), chrmaxes[i[0]])
    return(chrmaxes)

def get_chroffsets(chroms, chrmaxes, chromspace):
    """Calculate the offset for each chromosome based on the maximum value for that chromosome"""
    chroffsets = {}
    mainoffset = 0
    for i in chroms:
        chroffsets[i] = mainoffset
        mainoffset = int(mainoffset) + int(chrmaxes[i]) + int(chromspace)
    return(chroffsets)

# Both of these function just calculate offsets for plotting
def get_start_offset(row, chroffsets):
    return(chroffsets[row['chrom']] + int(row['start']))

def get_end_offset(row, chroffsets):
    return(chroffsets[row['chrom']] + int(row['end']))

def get_color_col(catcol):
    """Generate a color column for plotting. No longer used."""
    categories = np.unique(catcol)
    built_in_colors = [ "#1b9e77", "#d95f02", "#7570b3", "#e7298a", "#66a61e"]
    colors = built_in_colors[:len(categories)]
    #colors = np.linspace(0, 1, len(categories))
    colordict = dict(zip(categories, colors))

    colorcol = catcol.apply(lambda x: colordict[x])
    return colorcol

def parse_all_data(inconns, chromspace, names):
    """Parse input files, return a list with one data frame per input file"""
    alldatas = []
    if names:
        letters = names
    else:
        letters = string.ascii_lowercase + string.ascii_uppercase
    for index, inconn in enumerate(inconns):
        alldata = pd.read_csv(inconn, sep="\t", header=0)
        inconn.close()

        chrlist = alldata["chrom"].tolist()
        endlist = alldata["end"].tolist()

        chroms = sorted(list(set([str(x) for x in chrlist])))
        chrmaxes = get_chrmaxes(chrlist, endlist)
        chroffsets = get_chroffsets(chroms, chrmaxes, chromspace)

        alldata["start_offset"] = alldata.apply(lambda row: get_start_offset(row, chroffsets), axis=1)
        alldata["end_offset"] = alldata.apply(lambda row: get_end_offset(row, chroffsets), axis=1)
        alldata["cross"] = [letters[index] for x in range(len(alldata.index))]
        alldatas.append(alldata)
    return(alldatas)

def main():

    # parse all arguments
    parser = argparse.ArgumentParser("Visualize Hi-C pairing rates as a 2-d line plot.")
    parser.add_argument("input", nargs='*', help="Input file(s) generated by pairviz (default = stdin).")
    parser.add_argument("-o", "--output", help="output path (default = out.pdf).")
    parser.add_argument("-t", "--title", help="Title of plot (default = \"Pairing Rate\").")
    parser.add_argument("-p", "--proportion", help="If included, plot as a proportion of total reads in the region, rather than absolute (default = False).", action="store_true")
    parser.add_argument("-s", "--self", help="Also plot self-interactions (default = False).", action="store_true")
    parser.add_argument("-c", "--chromspace", help="bp of space to put between chromosomes in plot (default = 5000000).")
    parser.add_argument("-l", "--log", help="Log-scale the y-axis (default = False).", action="store_true")
    parser.add_argument("-i", "--stdin", help="take input from stdin along with other inputs.", action="store_true")
    parser.add_argument("-n", "--names", help="Names to use for plotting of input files (default = alphabetical letters).")

    args = parser.parse_args()

    # Set all variables to defaule values, handle argument logic
    output = "out.pdf"
    inconns = []
    if args.stdin or not input:
        inconns.append(sys.stdin)
    title = "Pairing Rate"
    proportion = False
    self = False
    chromspace = 5000000
    log = False
    pdf = False
    names = None
    if args.output:
        output = args.output
        pdf = True
    if args.input:
        for i in args.input:
            inconns.append(open(i, "r"))
    if args.title:
        title = args.title
    if args.proportion:
        proportion = args.proportion
    if args.self:
        self = args.self
    if args.chromspace:
        chromspace = args.chromspace
    if args.log:
        log = args.log
    if args.names:
        names = args.names.split(',')

    # set proportion vs total hits
    if proportion:
        my_y = 'pair_prop'
        alt_y = 'alt_prop'
    else:
        my_y = 'hits'
        alt_y = 'alt_hits'
    
    # make the combined data frame
    alldatas = parse_all_data(inconns, chromspace, names)
    big_alldata = pd.concat(alldatas, ignore_index=True)
    m_alldata = pd.melt(big_alldata, id_vars=['chrom', 'start', 'end', 'start_offset', 'end_offset', 'cross'])
    if self:
        mm_alldata = m_alldata[m_alldata.apply(lambda x: x['variable'] in (my_y, alt_y), axis=1)]
    else:
        mm_alldata = m_alldata[m_alldata.apply(lambda x: x['variable'] in (my_y), axis=1)]
    mm_alldata['value'] = mm_alldata['value'].astype(float)

    tempcols = [x for x in mm_alldata.columns]
    tempcols[3] = "Genome position (bp)"
    tempcols[-1] = "Hi-C contacts"
    mm_alldata.columns = [x for x in tempcols]

    # Plot it
    myplot=sns.relplot(x='Genome position (bp)', y='Hi-C contacts', data=mm_alldata, hue='cross', kind='line')
    if log:
        myplot.fig.get_axes()[0].set_yscale('log')
    
    # Save or show the plot
    if pdf:
        plt.savefig(output)
    else:
        plt.show()

if __name__ == "__main__":
    main()
#chrom   start   end     hit_type        alt_hit_type    hits    alt_hits        pair_prop       alt_prop        pair_totprop    pair_totgoodprop        pair_totcloseprop       winsize winstep
#2L      0       10001   paired  self    9       26780   0.00033595879   0.99966404      5.4952663e-08   6.5204028e-08   6.5352044e-08   10001   1000
#2L      1000    11001   paired  self    9       26780   0.00033595879   0.99966404      5.4952663e-08   6.5204028e-08   6.5352044e-08   10001   1000
#chrom
#start
#end
#hit_type
#alt_hit_type
#hits
#alt_hits
#pair_prop
#alt_prop
#pair_totprop
#pair_totgoodprop
#pair_totcloseprop
#winsize
#winstep
#
# start_offset
# end_offset
