# Pairviz

Pairviz is software for converting .pairs files that describe Hi-C data into linear measures of chromosome pairing. There are two working tools now, and more under construction:

## Requirements

Pairviz is written in Go and Python 3 and has the following requirements:

```
matplotlib
numpy
pandas
seaborn
```

### Associated repositories

- [https://github.com/jgbaldwinbrown/covplots](https://github.com/jgbaldwinbrown/covplots)
	- This package is designed for plotting data in .bed files and was designed for plotting the output of pairviz

## Working tools

### `pairviz`

Pairviz converts .pairs files to tab-separated tables that are ready for plotting. The usage is:

```
Usage of pairviz:
  -G	Print two entries for each chromosome location, one for each genome, correctly distinguishing self and paired reads (default = false).
  -c	Calculate whole-chromosome statistics, not sliding windows.
  -d int
    	Distance between two paired reads before they are ignored. (default -1)
  -f	Do not compute fpkm statistics.
  -g int
    	unused
  -i	Use Stdin as input (ignored; always do this anyway).
  -j	Output as JSON
  -m int
    	Minimum distance between two self reads reads. (default -1)
  -n string
    	Name to add to end of table.
  -pm int
    	Minimum distance between two paired reads. (default -1)
  -r string
    	Calculate statistics in a set of regions specified by this bedfile (not compatible with whole-chromosome statistics or window statistics).
  -rlen int
    	Length of reads in pairs (used to calculate overlapping or not; skipped otherwise). (default -1)
  -s int
    	Window step distance. (default -1)
  -sim int
    	Minimum distance between inward-facing self reads. (default -1)
  -w int
    	Window size. (default -1)
```

### `pairviz_plot.py`

Pairviz\_plot converts the tabular output of Pairviz into plots. Its usage is as follows:

```
usage: Visualize Hi-C pairing rates as a 2-d line plot. [-h] [-o OUTPUT]
                                                        [-t TITLE] [-p] [-f]
                                                        [-s] [-c CHROMSPACE]
                                                        [-l] [-i] [-n NAMES]
                                                        [-x X_AXIS_NAME]
                                                        [-y Y_AXIS_NAME]
                                                        [input [input ...]]

positional arguments:
  input                 Input file(s) generated by pairviz (default = stdin).

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        output path (default = out.pdf).
  -t TITLE, --title TITLE
                        Title of plot (default = "Pairing Rate").
  -p, --proportion      If included, plot as a proportion of total reads in
                        the region, rather than absolute (default = False).
  -f, --no\_fpkm         If included, plot FPKM rather than read counts
                        (default = False, can be combined with --proportion).
  -s, --self            Also plot self-interactions (default = False).
  -c CHROMSPACE, --chromspace CHROMSPACE
                        bp of space to put between chromosomes in plot
                        (default = 5000000).
  -l, --log             Log-scale the y-axis (default = False).
  -i, --stdin           take input from stdin along with other inputs.
  -n NAMES, --names NAMES
                        Names to use for plotting of input files (default =
                        alphabetical letters).
  -x X\_AXIS\_NAME, --x\_axis\_name X\_AXIS\_NAME
                        X axis name.
  -y Y\_AXIS\_NAME, --y\_axis\_name Y\_AXIS\_NAME
                        Y axis name.
```
